pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -0,0 +1,271 @@
+                        if (!partitionsSet.contains(existingPartition)) {
+                            futures.add(existingReader.closeAsync()
+                                    .thenRun(() -> readers.remove(existingPartition, existingReader)));
+                        }
+                    });
+
+                    return FutureUtil.waitForAll(futures)
+                            .thenRun(() -> schedulePartitionsCheck());
+                }).thenApply(__ -> this);
+    }
+
+    private void schedulePartitionsCheck() {
+        client.timer()
+                .newTimeout(this::checkForPartitionsChanges, conf.getAutoUpdatePartitionsSeconds(), TimeUnit.SECONDS);
+    }
+
+    private void checkForPartitionsChanges(Timeout timeout) {
+        if (timeout.isCancelled()) {
+            return ;
+        }
+
+        start().whenComplete((tw, ex) -> {
+           if (ex != null) {
+               log.warn("Failed to check for changes in number of partitions");
+           }
+        });
+    }
+
+    @Override
+    public int size() {
+        return data.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return data.isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(String key) {
+        return data.containsKey(key);
+    }
+
+    @Override
+    public T get(String key) {
+       return data.get(key);
+    }
+
+    @Override
+    public Set<Map.Entry<String, T>> entrySet() {
+       return data.entrySet();
+    }
+
+    @Override
+    public Set<String> keySet() {
+        return data.keySet();
+    }
+
+    @Override
+    public Collection<T> values() {
+        return data.values();
+    }
+
+    @Override
+    public void forEach(BiConsumer<String, T> action) {
+        data.forEach(action);
+    }
+
+    @Override
+    public void forEachAndListen(BiConsumer<String, T> action) {
+        // Ensure we iterate over all the existing entry _and_ start the listening from the exact next message
+        try {
+            listenersMutex.lock();
+
+            // Execute the action over existing entries
+            forEach(action);
+
+            listeners.add(action);
+        } finally {
+            listenersMutex.unlock();
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> closeAsync() {
+        return FutureUtil.waitForAll(
+                readers.values().stream()
+                        .map(Reader::closeAsync)
+                        .collect(Collectors.toList())
+        );
+    }
+
+    @Override
+    public void close() throws PulsarClientException {
+        try {
+            closeAsync().get();
+        } catch (Exception e) {
+            throw PulsarClientException.unwrap(e);
+        }
+    }
+
+    private void handleMessage(Message<T> msg) {
+        try {
+            if (msg.hasKey()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Applying message. key={} value={}", msg.getKey(), msg.getValue());
```

Comment:
It's better to print the topic name as well, we can easily to troubleshooting problems.

Reply:
@codelipenghui  

The topic name is logged during the initial replay:
https://github.com/apache/pulsar/pull/12838/files#diff-4c15ad9e44a19d82a9c40274cad3950b0af335cb40bd537b130300510a6857d8R251

Reply:
Yes, but we should do that also here, since you might have multiple tableview instances in the same client.